HTMLInputElement::setValueForUser should dispatch an input event

mcatanzaro opened the pull request:
#### fd13c500871d368b9f0b6d2f31a7ae57e4234e32
<pre>
HTMLInputElement::setValueForUser should dispatch an input event
<a href="https://bugs.webkit.org/show_bug.cgi?id=226023">https://bugs.webkit.org/show_bug.cgi?id=226023</a>
rdar://78571564

Reviewed by Aditya Keerthi.

HTMLInputElement::setValueForUser is used to autofill form elements. It
differs from setting the form value normally because it creates a change
event to trick websites into detecting user input. However, this is not
enough for some websites that only look for input events and not change
events to decide whether the user modified the form. On such websites,
the user has to enter an extra character and then delete it as a
workaround.

This fixes password autofill on <a href="https://my.cigna.com">https://my.cigna.com</a> and
<a href="https://myaccount.spireenergy.com">https://myaccount.spireenergy.com</a> in Epiphany.

* LayoutTests/fast/events/onchange-js-expected.txt:
* LayoutTests/fast/events/onchange-js.html: Add test to ensure change
  event is not sent until blur, since I&apos;ve removed that from the
  onchange-setvalueforuser.html test.
* LayoutTests/fast/forms/onchange-setvalueforuser.html: Change this test
  to not complain about the change event being sent before focus out.
* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement::setValueForUser): Fix the bug.
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::handleKeydownEvent): This seems to be using
setValueForUser by mistake. Call setValue directly instead to avoid
breaking fast/forms/search-cancel-button-events.html.
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::defaultEventHandler): Ditto.

Canonical link: <a href="https://commits.webkit.org/259434@main">https://commits.webkit.org/259434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/247c20ffe7a02e531769cac6e81588dbe2ecebd0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114051 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174254 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4778 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112967 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110532 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39115 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108237 "webkitpy-tests (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26215 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80779 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94682 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7204 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27573 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92661 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4944 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4155 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30216 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103583 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47125 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101351 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6502 "Build was cancelled. Recent messages:Cleaned up git repository; Checked out pull request; Verified commit is squashed; Reviewed by Aditya Keerthi; Validated commit message; compiling; layout-tests (exception)") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9090 "Build is being retried. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25290 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3456 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/onchange-setvalueforuser.html:
```diff
@@ -19,13 +19,6 @@
             testRunner.setValueForUser(tf, 'Hello!');
         }
 
-        // Should not fire the event until focus is lost.
```
The `change` event typically fires after `focus` is lost for text input types, so this seems like an incorrect behavior change.

See https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event.

mcatanzaro replied:
But we are just autofilling, so the element is never going to get focused in the first place. Does that really make sense here?

mcatanzaro replied:
Well it happens in this test because the test manually does a focus/unfocus before calling setValueForUser, but Epiphany (and presumably Safari?) does not do that.

pxlcoder replied:
Makes sense. This scenario also fits the description of "When the user commits the change explicitly...", in which case it's okay to fire a change event immediately.

How would you feel about updating `LayoutTests/fast/events/onchange-js.html` to assert that a change event is not fired before losing focus when typing – so that we don't lose the coverage that this test provided.

mcatanzaro replied:
Sure. I've attempted this. Let's see if the EWS likes it.

pxlcoder commented on Source/WebCore/html/SearchInputType.cpp:
```diff
@@ -146,7 +146,7 @@ auto SearchInputType::handleKeydownEvent(KeyboardEvent& event) -> ShouldCallBase
     if (key == "U+001B"_s) {
         Ref<HTMLInputElement> protectedInputElement(*element());
-        protectedInputElement->setValueForUser(emptyString());
+        protectedInputElement->setValue({ }, DispatchChangeEvent);
```
It seems strange to stop using `setValueForUser`, given that clearing the value here is a user-driven action. Also I think we need to fire an `input` event here as `input` events should fire whenever value changes occur.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event

mcatanzaro replied:
OK, I can switch it back and update the expectations of fast/forms/search-cancel-button-events.html instead.

Instead of:

```
Click the cancel button
PASS search.value is ""
PASS changeEventCounter is 0
PASS inputEventCounter is 1
```

We'll have:

```
Click the cancel button
PASS search.value is ""
PASS changeEventCounter is 1
PASS inputEventCounter is 1
```

0 looked better to me because the value there didn't really change. It's a little confusing because the code change looks like it's removing an input event dispatch, not a change event, and I don't know why that happens.

pxlcoder replied:
Actually, I think this change is fine as-is.

Here is the logic in `TextFieldInputType`:

```
    case DispatchChangeEvent:
        // If the user is still editing this field, dispatch an input event rather than a change event.
        // The change event will be dispatched when editing finishes.
        if (input->focused())
            input->dispatchFormControlInputEvent();
        else
            input->dispatchFormControlChangeEvent();
        break;
```

There is a change in the test case, it the value changes from "foo" to "".

Pressing the cancel button focuses the element, so `DispatchChangeEvent` will fire an `input` event, but not a `change` event immediately. The `change` event will then get fired when the element is blurred.

Our current behavior also matches Chrome.

mcatanzaro replied:
Uh, I might need to investigate this more, because it seems the test is still failing even though what I have now should be equivalent to what was there before.

pxlcoder replied:
`{ }` is a null string, whereas there was an empty string before – I think your latest change should work.

mcatanzaro replied:
Huh... that's a kind of important difference. I probably should have learned that sooner than now.

mcatanzaro replied:
It didn't work. O_O

mcatanzaro replied:
Which is weird, because it sure looks like these calls should now be completely identical. We must be missing something.

mcatanzaro replied:
Ahhh, I missed a spot! `SearchFieldCancelButtonElement::defaultEventHandler` in TextControlInnerElements.cpp. That must be it.

mcatanzaro commented on LayoutTests/fast/forms/onchange-setvalueforuser.html:
```diff
@@ -19,13 +19,6 @@
-            return;
-        }
-
-        tf.blur();
```
Oops, I forgot to remove the corresponding `tf.focus()` up above.

pxlcoder reviewed (approved):
Thanks for looking at this!

mcatanzaro committed:
HTMLInputElement::setValueForUser should dispatch an input event
https://bugs.webkit.org/show_bug.cgi?id=226023
rdar://78571564

Reviewed by Aditya Keerthi.

HTMLInputElement::setValueForUser is used to autofill form elements. It
differs from setting the form value normally because it creates a change
event to trick websites into detecting user input. However, this is not
enough for some websites that only look for input events and not change
events to decide whether the user modified the form. On such websites,
the user has to enter an extra character and then delete it as a
workaround.

This fixes password autofill on https://my.cigna.com and
https://myaccount.spireenergy.com in Epiphany.

* LayoutTests/fast/events/onchange-js-expected.txt:
* LayoutTests/fast/events/onchange-js.html: Add test to ensure change
  event is not sent until blur, since I've removed that from the
  onchange-setvalueforuser.html test.
* LayoutTests/fast/forms/onchange-setvalueforuser.html: Change this test
  to not complain about the change event being sent before focus out.
* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement::setValueForUser): Fix the bug.
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::handleKeydownEvent): This seems to be using
setValueForUser by mistake. Call setValue directly instead to avoid
breaking fast/forms/search-cancel-button-events.html.
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::defaultEventHandler): Ditto.

Canonical link: https://commits.webkit.org/259434@main
